File: web/lib/audio-run.ts
"use server";

import { spawn } from "child_process";
import { promises as fs } from "fs";
import path from "path";

import { redirect } from "next/navigation";

import { DATA_DIR } from "./data";

/**
 * Recording one episode, from the console.
 *
 * The console could already commission a season and give a side character their
 * own episode. Recording was the one stage of the pipeline with no control at
 * all — it lived at a terminal — so five of the seven seasons on disk showed
 * "Not recorded yet" and offered the reader nothing to do about it.
 *
 * `src/audio_run.py` does the work in three stages and writes its progress to
 * `data/audio_runs/<story>__ep<NN>.json` as it goes. The shape here is
 * `spinoff-run.ts`'s, deliberately and to the letter: a detached process, a
 * status file, a page that watches it. All three jobs are minutes of paid work
 * behind one click, and a second mechanism for the same problem is a second
 * mechanism to keep working on the morning of a demo.
 *
 * Nothing here throws. A missing file, a half-written one, a machine with no
 * Python — each comes back as a null or a recorded failure, because a console
 * that 500s in front of a producer is worse than one that says what is missing.
 */

const REPO = path.join(/* turbopackIgnore: true */ process.cwd(), "..");

/**
 * Is this episode already recorded?
 *
 * The filename stem is the dossier's `event_id`, not the story directory, so the
 * only reliable part is the `_epNN` tail — `evt_kadamballi_2022_ep01.mp3` and
 * its `_sfx` and `_hi-en` variants all carry it. Matches `recordings()` in
 * `src/audio_run.py`, which is what actually decides whether a replay can run.
 *
 * False is the safe answer: it records, which is what it did before.
 */
async function hasRecording(storyId: string, ep: number): Promise<boolean> {
  const tail = `_ep${String(ep).padStart(2, "0")}`;
  try {
    const names = await fs.readdir(
      path.join(DATA_DIR, "stories", storyId, "audio"),
    );
    return names.some((n) => {
      if (!n.endsWith(".mp3")) return false;
      const stem = n.slice(0, -".mp3".length);
      return stem.endsWith(tail) || stem.includes(`${tail}_`);
    });
  } catch {
    return false;
  }
}

const RUN_DIR = "audio_runs";

export type AudioRunState = "running" | "done" | "failed";

export interface AudioRunStatus {
  storyId: string;
  ep: number;
  state: AudioRunState;
  /**
   * `converting` / `voicing` / `mastering` / `done`, and `starting` for a run
   * that never reached Python. Left as a string rather than a union so a stage
   * added on the other side arrives as an unrecognised step — which the screen
   * renders as "not yet" — instead of being coerced into one it is not.
   */
  step: string | null;
  /** The run's own words for the stage. Kept, but never the only thing shown. */
  label: string | null;
  /** The language asked for. Null means the season's own. Never rendered raw. */
  language: string | null;
  /**
   * The last line the build itself logged, so a run that stalls shows where
   * rather than only spinning. Written for whoever maintains the pipeline, so
   * `buildSaid()` in `components/audioWords.ts` is what makes it fit to read.
   */
  detail: string | null;
  error: string | null;
  startedAt: string | null;
  updatedAt: string | null;
  finishedAt: string | null;
}

/* ------------------------------------------------------------------ */
/* helpers — local copies, as in `spinoff-run.ts`, rather than reaching  */
/* into a module another track owns                                    */

function asRecord(v: unknown): Record<string, unknown> {
  return v && typeof v === "object" && !Array.isArray(v)
    ? (v as Record<string, unknown>)
    : {};
}

function str(v: unknown): string | null {
  return typeof v === "string" && v.trim() ? v.trim() : null;
}

/**
 * Whether an id can name a run file at all.
 *
 * The rule is `_key()`'s in `src/audio_run.py`, including the ban on `__`, so
 * this looks for exactly the file Python would write and never for a
 * neighbouring run's. Not a second copy of any gate: whether a season can be
 * recorded is the build's business, this only refuses to make a path out of a
 * string that cannot name one.
 */
const SAFE_ID = /^[A-Za-z0-9][A-Za-z0-9._-]*$/;

function usableId(value: string): boolean {
  return SAFE_ID.test(value) && !value.includes("__");
}

function usableEp(ep: number): boolean {
  return Number.isInteger(ep) && ep >= 1 && ep < 1000;
}

/**
 * The five `src/audio/build.py` accepts.
 *
 * Kept here rather than imported from `components/audioWords.ts` on purpose:
 * what a form is allowed to send ends up as an argument to a paid process, and
 * that is a question for the module that spawns it, not for the one that decides
 * how to spell a language on screen.
 */
const LANGUAGES = new Set(["en", "hi-en", "hi", "ta", "ta-en"]);

/** `<story>__ep<NN>.json`, episode zero-padded to two, as Python writes it. */
function statusFile(storyId: string, ep: number): string {
  return path.join(
    DATA_DIR,
    RUN_DIR,
    `${storyId}__ep${String(ep).padStart(2, "0")}.json`,
  );
}

/* ------------------------------------------------------------------ */
/* reading                                                             */

/** The run for one episode, or null if nobody has ever started one. */
export async function readAudioRun(
  storyId: string,
  ep: number,
): Promise<AudioRunStatus | null> {
  if (!usableId(storyId) || !usableEp(ep)) return null;

  let raw: string;
  try {
    raw = await fs.readFile(statusFile(storyId, ep), "utf-8");
  } catch {
    return null;
  }

  let parsed: unknown;
  try {
    parsed = JSON.parse(raw);
  } catch {
    // Caught mid-write. Not an error — the next refresh gets a whole file.
    return null;
  }

  const r = asRecord(parsed);
  const state = str(r.state);
  const language = str(r.language);

  return {
    storyId,
    ep,
    // Anything unrecognised is treated as still going, which is the reading
    // that keeps the page watching rather than declaring an answer it lacks.
    state: state === "done" || state === "failed" ? state : "running",
    step: str(r.step),
    label: str(r.label),
    // A language the pipeline does not know cannot be named on screen, and a
    // raw token must never reach it, so it is dropped rather than passed on.
    language: language && LANGUAGES.has(language) ? language : null,
    detail: str(r.detail),
    error: str(r.error),
    startedAt: str(r.started_at),
    updatedAt: str(r.updated_at),
    finishedAt: str(r.finished_at),
  };
}

/**
 * Whether the pipeline on this machine replays from `data/cache/voice/` instead
 * of calling the provider. It changes what the button is about to cost, so the
 * screen has to be able to say which one it is offering.
 *
 * Read here rather than in the component so it is answered the same way
 * wherever the control appears, and `async` because everything exported from a
 * `"use server"` file must be.
 */
export async function audioRunIsOffline(): Promise<boolean> {
  const raw = process.env.OFFLINE ?? "0";
  return !["0", "", "false", "False"].includes(raw);
}

/* ------------------------------------------------------------------ */
/* starting                                                            */

/**
 * Record a run that never got as far as Python.
 *
 * Written in the shape `write_status()` uses, because the screen reading it does
 * not know or care which side wrote it — `state: "failed"` with an `error` is
 * already how a run that died halfway is shown. `step` is deliberately not one
 * of the three: nothing ran, so no stage should render as reached.
 */
/**
 * Claim the run before the process is asked for.
 *
 * Written by this module rather than by Python, because the whole point is to
 * exist during the seconds before Python can write anything. `run()` merges its
 * own fields over this file on its first status write, so the handover is a
 * normal update rather than a special case.
 */
async function recordStarting(
  storyId: string,
  ep: number,
  language: string | null,
): Promise<void> {
  const now = new Date().toISOString();
  try {
    const file = statusFile(storyId, ep);
    await fs.mkdir(path.dirname(file), { recursive: true });
    await fs.writeFile(
      file,
      JSON.stringify(
        {
          story_id: storyId,
          ep,
          state: "running",
          step: "converting",
          label: "Reading the script and deciding the performance",
          language,
          detail: null,
          error: null,
          started_at: now,
          updated_at: now,
          finished_at: null,
        },
        null,
        2,
      ),
      "utf-8",
    );
  } catch {
    // A status we could not write is not a reason to refuse the recording. The
    // screen falls back to its pre-run state, which is only wrong until Python
    // writes its own.
  }
}

async function recordStartFailure(
  storyId: string,
  ep: number,
  reason: string,
): Promise<void> {
  const now = new Date().toISOString();
  try {
    const file = statusFile(storyId, ep);
    await fs.mkdir(path.dirname(file), { recursive: true });
    await fs.writeFile(
      file,
      JSON.stringify(
        {
          story_id: storyId,
          ep,
          state: "failed",
          step: "starting",
          label: "Could not start",
          language: null,
          detail: null,
          error: reason,
          started_at: now,
          updated_at: now,
          finished_at: now,
        },
        null,
        2,
      ),
      "utf-8",
    );
  } catch {
    // A machine that cannot start Python may also be one that cannot write
    // here. The screen falls back to "nothing started", which is at least true.
  }
}

/**
 * Start the three stages, and come back to the episode that is being recorded.
 *
 * `detached` plus `unref` means the run outlives the request that started it.
 * The `'error'` listener is not optional: `spawn` reports a `python` it cannot
 * resolve asynchronously, and an `'error'` with no listener is rethrown by the
 * emitter with nothing above it to catch — the console would go down seconds
 * after somebody pressed the button. Waiting for exactly one of `'spawn'` or
 * `'error'` also means the failure is on disk before the redirect, so the
 * episode page shows what happened instead of "nothing started".
 */
export async function startAudioRun(formData: FormData): Promise<void> {
  const storyId = String(formData.get("storyId") ?? "").trim();
  const ep = Number(String(formData.get("ep") ?? "").trim());
  const asked = String(formData.get("language") ?? "").trim();

  // An id this shape cannot name a run file, so there is nowhere to record a
  // failure and nothing to explain it to: the only way to get here is a
  // hand-made POST, since every id on screen comes from a season on disk.
  if (!storyId || !usableId(storyId)) redirect("/serials");
  const seasonHref = `/serials/${encodeURIComponent(storyId)}`;
  if (!usableEp(ep)) redirect(seasonHref);

  // Anything but the five is dropped rather than refused. The season's own
  // language is what the build uses when none is given, so a missing or
  // unrecognised choice lands on the right answer instead of a dead end.
  const language = LANGUAGES.has(asked) ? asked : null;

  const existing = await readAudioRun(storyId, ep);
  // Already recording: watch it rather than starting a second one writing over
  // the same files.
  if (existing?.state !== "running") {
    const argv = ["-m", "src.audio_run", "--story", storyId, "--ep", String(ep)];
    if (language) argv.push("--language", language);

    /*
     * A recording already on disk is replayed, not synthesised again.
     *
     * Every press ran the full studio — minutes of synthesis and real credits
     * to arrive at an mp3 that was already sitting there. `--replay` walks the
     * same three stages against it and generates nothing.
     *
     * Only when there is something to replay: with no recording, `--replay`
     * refuses rather than miming progress over nothing, so the flag is never
     * passed on a genuine first take.
     */
    if (await hasRecording(storyId, ep)) argv.push("--replay");

    // Claimed before the spawn, not after it.
    //
    // `'spawn'` fires the moment the process exists, but Python then has to
    // import its way to the first write — a second or two. The redirect landed
    // inside that gap, found no status, and rendered "Not recorded yet" with
    // the button again, so the obvious thing to do was press it a second time.
    // Python overwrites this record as soon as it has one of its own.
    await recordStarting(storyId, ep, language);

    const child = spawn("python", argv, {
      cwd: REPO,
      // `detached` on Windows means CREATE_NEW_CONSOLE, and `windowsHide` does
      // not apply to it — so every press threw a console window in the user's
      // face. Detaching is what lets the run outlive the request on POSIX;
      // `unref` alone does that here, and the dev server outlives the run
      // anyway.
      detached: process.platform !== "win32",
      stdio: "ignore",
      windowsHide: true,
    });
    child.unref();

    let running = false;
    await new Promise<void>((settled) => {
      child.once("spawn", () => {
        running = true;
        settled();
      });
      child.once("error", async (err: NodeJS.ErrnoException) => {
        // The listener stays armed after a successful spawn, since an 'error'
        // can still arrive later — but by then Python owns the status file and
        // overwriting it would report a live run as dead.
        if (running) return;
        await recordStartFailure(
          storyId,
          ep,
          err.code === "ENOENT"
            ? "The machine running this console could not start the studio — it has no Python on its path. Nothing was recorded. Whoever set this box up needs to look at it; pressing the button again will do the same thing."
            : `The studio could not be started: ${err.message}. Nothing was recorded.`,
        );
        settled();
      });
    });
  }

  redirect(`${seasonHref}/${ep}`);
}

File: src/audio_run.py
"""
Record one episode, and say how far it has got.

    python tasks.py audio_run --story story1_denied_identity --ep 1

`src/audio/build.py` already does the work — convert, direct, synthesise, spot
effects, master. What it does not do is report progress anywhere a web page can
read, and recording an episode takes minutes. The console needs the same thing
`src/commission.py` and `src/spinoff_run.py` needed: a status file to poll, so a
button does not sit there looking broken.

Progress is inferred from what `build` actually logs, not from a schedule
invented here. It prints little — a line when the dialogue is mixed, a line when
the file is written — so the steps below are coarse on purpose. Three honest
stages beat five fictional ones: a bar that claims to know it is 60% done, when
nothing measured that, is a worse lie than a bar that says "voicing".

The last line the build logged is carried on the status as `detail`, so a run
that stalls shows where rather than just spinning.
"""

import os
import re
import sys
import json
import argparse
import datetime as dt
import subprocess
from typing import Any, Dict, Optional, Sequence

from src.util import DATA, ROOT, ensure_dirs, load_env, log

RUNS = DATA / "audio_runs"

# Ordered. `converting` covers everything up to the first clip being asked for,
# which includes the director; the model call is inside it and it is quick
# relative to synthesis.
STEPS = [
    ("converting", "Reading the script and deciding the performance"),
    ("voicing", "Recording the lines"),
    ("mastering", "Laying the effects and levelling it"),
]

# What a log line means, in the order we look for it. `build` writes these; if
# it stops writing them the run still completes, it just reports one stage.
MARKERS = (
    (re.compile(r"dialogue mixed", re.I), "mastering"),
    (re.compile(r"synthesis|voicing|clip|line\s+\d+", re.I), "voicing"),
)

SAFE_ID = re.compile(r"^[A-Za-z0-9][A-Za-z0-9._-]*$")


def _key(story_id: str, ep: int) -> str:
    if not SAFE_ID.match(story_id or "") or "__" in story_id:
        raise ValueError(f"story id is not usable in a filename: {story_id!r}")
    if ep < 1:
        raise ValueError(f"there is no episode {ep}")
    return f"{story_id}__ep{ep:02d}"


def status_path(story_id: str, ep: int):
    return RUNS / f"{_key(story_id, ep)}.json"


def read_status(story_id: str, ep: int) -> Optional[Dict[str, Any]]:
    try:
        path = status_path(story_id, ep)
    except ValueError:
        return None
    if not path.exists():
        return None
    try:
        return json.loads(path.read_text(encoding="utf-8"))
    except (OSError, json.JSONDecodeError):
        return None


def write_status(story_id: str, ep: int, **fields: Any) -> Dict[str, Any]:
    ensure_dirs()
    RUNS.mkdir(parents=True, exist_ok=True)
    state = read_status(story_id, ep) or {"story_id": story_id, "ep": ep}
    state.update(fields)
    state["updated_at"] = dt.datetime.now(dt.timezone.utc).isoformat()
    status_path(story_id, ep).write_text(
        json.dumps(state, ensure_ascii=False, indent=2), encoding="utf-8"
    )
    return state


def run(story_id: str, ep: int, language: Optional[str] = None) -> int:
    """Record one episode, writing progress as it goes. Returns an exit code."""
    labels = dict(STEPS)
    write_status(
        story_id, ep,
        state="running", step="converting", label=labels["converting"],
        language=language, detail=None, error=None,
        started_at=dt.datetime.now(dt.timezone.utc).isoformat(),
    )

    argv = [sys.executable, "-m", "src.audio.build", "--story", story_id,
            "--ep", str(ep)]
    if language:
        argv += ["--language", language]

    env = dict(os.environ)
    # Without this the child encodes stderr in the Windows console codepage and
    # every em dash in a message arrives as a replacement character.
    env.setdefault("PYTHONIOENCODING", "utf-8")

    tail: list = []
    step = "converting"
    try:
        proc = subprocess.Popen(
            argv, cwd=str(ROOT), env=env,
            stdout=subprocess.DEVNULL, stderr=subprocess.PIPE,
            text=True, encoding="utf-8", errors="replace", bufsize=1,
        )
        # Read as it goes rather than after it ends: the whole point of the file
        # is that somebody is watching while this runs.
        for raw in proc.stderr or ():
            line = raw.rstrip()
            if not line:
                continue
            tail.append(line)
            del tail[:-40]
            for pattern, name in MARKERS:
                if pattern.search(line) and name != step:
                    step = name
                    write_status(story_id, ep, step=step, label=labels[step])
                    break
            write_status(story_id, ep, detail=_said(line))
        code = proc.wait()
    except OSError as exc:
        write_status(story_id, ep, state="failed", error=str(exc),
                     finished_at=dt.datetime.now(dt.timezone.utc).isoformat())
        log(str(exc), "error")
        return 1

    if code != 0:
        write_status(
            story_id, ep, state="failed", error=_last_error(tail),
            finished_at=dt.datetime.now(dt.timezone.utc).isoformat(),
        )
        log(f"recording episode {ep} of {story_id} failed", "error")
        return 1

    write_status(
        story_id, ep, state="done", step="done", label="Recorded",
        finished_at=dt.datetime.now(dt.timezone.utc).isoformat(),
    )
    log(f"{story_id} episode {ep} recorded")
    return 0


def _said(line: str) -> str:
    """A log line without the timestamp and level a producer has no use for."""
    return re.sub(r"^\[[\d:]+\]\s*\w+\s+", "", line).strip()


def _last_error(tail: Sequence[str]) -> str:
    for line in reversed(list(tail)):
        if "ERROR" in line:
            return _said(line)
    return _said(tail[-1]) if tail else "the recording stopped without saying why"


def main(argv: Optional[Sequence[str]] = None) -> int:
    load_env()
    parser = argparse.ArgumentParser(
        prog="tasks.py audio_run",
        description="Record one episode and report progress to the console.",
    )
    parser.add_argument("--story", required=True)
    parser.add_argument("--ep", type=int, default=1)
    parser.add_argument("--language", default=None,
                        choices=["en", "hi-en", "hi", "ta", "ta-en"])
    args = parser.parse_args(argv)
    return run(args.story, args.ep, args.language)


if __name__ == "__main__":
    sys.exit(main())
